synapse/tests/test_lib_net.py

```diff
@@ -54,3 +244,171 @@ def onlink(link):
+
+                def linked(self):
+                    steps.done('lisn')
+
+            class ConnLink(s_net.Link):
+
+                def linked(self):
+                    steps.done('conn')
+                    self.tx(('wat', {}))
+
+            with s_net.Plex() as plex:
+
+                def onconn(ok, link):
+                    conn = ConnLink(link)
+                    link.onrx(conn.rx)
+                    conn.linked()
+
+                def onlink(link):
+                    lisn = LisnLink(link)
+                    link.onrx(lisn.rx)
+                    lisn.linked()
+
+                addr = plex.listen(('127.0.0.1', 0), onlink)
+                plex.connect(addr, onconn)
+
+                steps.waitall(timeout=2)
+                self.true(stream.wait(10))
+
+            stream.seek(0)
+            mesgs = stream.read()
+            self.isin(expected_msg, mesgs)
+
+    def test_lib_net_finis(self):
+
+        names = ('conn', 'lisn')
+        steps = self.getTestSteps(names)
+        expected_msg = 'I WAS FINID'
+        with self.getLoggerStream('synapse.tests.test_lib_net', expected_msg) as stream:
+
+            class LisnLink(s_net.Link):
+
+                def linked(self):
+                    steps.done('lisn')
+
+            class ConnLink(s_net.Link):
+
+                def linked(self):
+                    self.rxfini()
+                    self.txfini(data='GOODBYE')
+                    self.txfini()
+                    steps.done('conn')
+
+                def onfini(self, msg):
+                    logger.error('I WAS FINID')
+
+            with s_net.Plex() as plex:
+
+                def onconn(ok, link):
+                    conn = ConnLink(link)
+                    link.onrx(conn.rx)
+                    conn.linked()
+
+                def onlink(link):
+                    lisn = LisnLink(link)
+                    link.onrx(lisn.rx)
+                    lisn.linked()
+
+                addr = plex.listen(('127.0.0.1', 0), onlink)
+                plex.connect(addr, onconn)
+
+                steps.waitall(timeout=2)
+                self.true(stream.wait(10))
+
+            stream.seek(0)
+            mesgs = stream.read()
+            self.isin(expected_msg, mesgs)
+
+    def test_lib_net_connectfail(self):
+        expected_msg = 'connect() onconn failed'
+        with self.getLoggerStream('synapse.lib.net', expected_msg) as stream:
+            with s_net.Plex() as plex:
+                plex.connect(('127.0.0.1', 0), None)
+                self.true(stream.wait(10))
+
+            stream.seek(0)
+            mesgs = stream.read()
+            self.isin(expected_msg, mesgs)
+
+    def test_lib_net_listenfail(self):
+        expected_msg = 'listen() onlink error'
+        with self.getLoggerStream('synapse.lib.net', expected_msg) as stream:
+
+            class LisnLink(s_net.Link):
+
+                def linked(self):
+                    steps.done('lisn')
```

Comment:
``steps`` is not defined

Reply:
This whole class can actually be removed

Reply:
fixed